### ISSUE TITLE
Consider using Changesets for versioning and generating CHANGELOG

### ISSUE THREAD

**Issue by taras** [spotify/backstage#1839]
## Context

I came across this project recently while researching developer portals for a large platform project. I've been reading your RFCs and digging into the architecture. We're now planning a spike to see how it'll fit our use case. One of our considerations is whether your roadmap aligns with ours. We would love to be able to leverage your plans and contribute to them. 

To that end, it would be helpful for me to be able to follow along with your progress via a CHANGELOG. I couldn't find one so I wanted to make a suggestion based on my experience maintaining large monorepos in [BigTest](https://github.com/thefrontside/bigtest) and [effection.js](https://github.com/thefrontside/effection). 

If you like the suggestion, I can implement it.

## Feature Suggestion

Have you considered using [Changesets](https://github.com/atlassian/changesets) to manage versioning and generating CHANGELOGS?

We've been using them for a while and found the following benefits,

1. Distribute the responsibility of preparing releases amongst all contributors who make up a release.
2. Automatically determine impact of the release based on all accumulated changes for each package.
3. Ensure that versions of each dependency are properly updated.

The result is this,

1. A release is always 1 merge button away (there is no extra work to prepare a release)
2. Every package gets a CHANGELOG automatically
3. You have fewer broken releases where dependencies are not properly being installed

## Possible Implementation

There are 4 key components to Changesets.

1. Directory of changesets
2. Changeset check
3. Changeset GitHub Action
4. Synchronize NPM with master 

### 1. Directory of changesets

`.changeset` the directory is where changesets get added. 

A change is a markdown file that has the following structure:

```md
---
"@bigtest/bundler": minor
"@bigtest/cli": patch
---
set process.env.NODE_ENV='production' inside the test bundle. This
will ensure that 3rd party packages that depend on this will continue
to function
```

The frontmatter header describes the impact of changes. The body will be used to describe the change. You can have a bunch of these files in the changeset directory. These files can be generated by hand, with a link from a changeset comment or with a [CLI](https://www.npmjs.com/package/@changesets/cli). 

<img width="941" alt="image" src="https://user-images.githubusercontent.com/74687/89460833-ef3b4200-d738-11ea-87a0-d7f86e60411c.png">

### 2. Changeset bot & check

Changeset Bot has to be added to the project. It creates comments like the one above that describes changesets that are added to the PR. It adds an optional check that you can use to prevent a PR from being merged without a changeset. This forces contributors to describe their changes before the PR is merged and ensures that releases are always ready.

### 3. Changeset GitHub Action

This action runs on master. When the action finds changesets in master, it will automatically create a [Version Packages PR](https://github.com/thefrontside/bigtest/pull/429). This PR will remove changesets and update versions of affected packages. When a new commit is pushed into master, Version Packages PR gets updated. This PR includes autogenerated CHANGELOGS for each package.

### 4. Synchronize NPM with master 

I believe you already have this functionality. When you commit is added to master that has unpublished versions of packages, automatically publish these packages to NPM.

---

Let me know if you have any questions about this. I'm happy to clarify this process. If you like, we can set it up for you.

**Comment by Rugvip** [comment]
Thank you for this suggestion! Loving especially the idea of #**3**. We do indeed have #**4** in place already.

As this is a pretty long-term decision we want to have a look at some alternatives as well, but it's definitely something we've been wanting to add and this looks like a good option.

We currently keep the versions of all packages in the repo in sync. I don't think we'll want to do that long-term, but it makes sense for now. Do you know if the Changeset tooling is compatible with our versioning strategy as well?

Another option is https://conventionalcommits.org, do you have any experience using that and know how it compares?

**Comment by taras** [comment]
> We currently keep the versions of all packages in the repo in sync. Do you know if the Changeset tooling is compatible with our versioning strategy as well?

They have a mechanism called [Linked Packages](https://github.com/atlassian/changesets/blob/master/docs/linked-packages.md) that allows to declare which packages have to versioned together in the config file. You can do lock step versioning by making all packages linked. 

> Another option is https://conventionalcommits.org, do you have any experience using that and know how it compares?

We used similar approach on a few projects but found it lacking with monorepoes. If your PR introduces changes to more than 1 package than you have to make sure that each package gets it's own commit. If you squash commits, you'll loose the association. 

Enforcing Conventional Commits requires merging through additional tooling because GitHub Merge UI doesn't have a way to enforce the format. Then you get into linting these commits and it just adds a lot of manual work to enforce the titles of these commits.

**Comment by andrewthauer** [comment]
I cannot comment on `changesets`, but it does look pretty comprehensive. However, I have used the [conventional-changelog[(https://github.com/conventional-changelog/conventional-changelog) packages for a couple of monorepos and have been pretty happy with it. There is tooling for linting commit messages, publishing, tagging, and creating GitHub releases. Once the tooling is setup you can mostly rely on the commit messages to drive everything (including determining the appropriate semantic version bump). It can be pretty seamless and easy from a release process perspective. There are some large open source monorepo based projects using `conventional commits` approach successfully (most notably `angular`).

**Comment by Rugvip** [comment]
Hi @taras! After a lot of deliberation we've arrived at wanting to move forward with Changesets. :grin:

I know it was a long time since you opened the issue, but if you're still happy to help setting it up that would be awesome!

We're also planning to move away from our `.alpha` prereleases, and start properly using semver instead, although we still want to keep versions between all packages in sync. At least for now.

**Comment by taras** [comment]
@Rugvip cool :) I'll be happy get this setup for you. I'll get my mind back into it later this week.

**Comment by taras** [comment]
@Rugvip I'm finishing my initial deployment of Backstage today. Once I do, I'll work on this. Probably later this week.

**Comment by taras** [comment]
https://github.com/release-drafter/release-drafter is interesting because it creates a GitHub release. We might want to create a release when we actually publish versions of the package. I'll look into this as a follow-up.

**Comment by taras** [comment]
Closed by #2437
